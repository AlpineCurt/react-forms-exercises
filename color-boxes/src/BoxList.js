import React, { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import {v4 as uuid} from "uuid";

const BoxList = () => {
    const INITIAL_STATE = [
        {key: uuid(), width: "200px", height: "100px", backgroundColor: "aqua"},
        {key: uuid(), width: "400px", height: "75px", backgroundColor: "peachpuff"}
    ]
    const [boxes, setBoxes] = useState([]);
    const addBox = (newBox) => {
        setBoxes(boxes => [...boxes, {
                                        key: uuid(),
                                        width: `${newBox.width}px`,
                                        height: `${newBox.height}px`,
                                        backgroundColor: newBox.backgroundColor
                                    }]);
    }
    const removeBox = (id) => {
        setBoxes(boxes.filter((box) => box.key !== id));
    }
    return (
        <>
            <NewBoxForm addBox={addBox}/>
            <div className="BoxList">
                {boxes.map(({key, width, height, backgroundColor}) => (
                    <Box
                        id={key}
                        key={key}
                        width={width}
                        height={height}
                        backgroundColor={backgroundColor}
                        handleRemoveBox={removeBox}
                    />
                ))}
            </div>
        </>
    )
}

export default BoxList;