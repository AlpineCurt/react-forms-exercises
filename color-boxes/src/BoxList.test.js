import React from "react";
import { render, fireEvent } from "@testing-library/react"
import BoxList from "./BoxList";

test("renders without crashing", () => {
    render(<BoxList />);
});

test("matches snapshot", () => {
    const {asFragment} = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

test("add new box", () => {
    const {getByLabelText, queryByText} = render(<BoxList />);
    const widthInput = getByLabelText("Width");
    const heightInput = getByLabelText("Height");
    const colorInput = getByLabelText("Color");
    const btn = queryByText("Add Box");
    expect(queryByText("X")).not.toBeInTheDocument();
    fireEvent.change(widthInput, {target: {value: "222"}});
    fireEvent.change(heightInput, {target: {value: "111"}});
    fireEvent.change(colorInput, {target: {value: "purple"}});
    fireEvent.click(btn);
    expect(queryByText("X")).toBeInTheDocument();
});

test("remove a box", () => {
    const {getByLabelText, queryByText} = render(<BoxList />);
    const widthInput = getByLabelText("Width");
    const heightInput = getByLabelText("Height");
    const colorInput = getByLabelText("Color");
    const btn = queryByText("Add Box");
    expect(queryByText("X")).not.toBeInTheDocument();
    fireEvent.change(widthInput, {target: {value: "222"}});
    fireEvent.change(heightInput, {target: {value: "111"}});
    fireEvent.change(colorInput, {target: {value: "purple"}});
    fireEvent.click(btn);
    expect(queryByText("X")).toBeInTheDocument();
    const removeBtn = queryByText("X");
    fireEvent.click(removeBtn);
    expect(queryByText("X")).not.toBeInTheDocument();
})