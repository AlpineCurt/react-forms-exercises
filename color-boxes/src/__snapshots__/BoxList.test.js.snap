// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Add a Box
    </h3>
    <form>
      <label
        for="form-width"
      >
        Width
      </label>
      <input
        id="form-width"
        name="width"
        placeholder="Width"
        type="text"
        value=""
      />
      <label
        for="form-height"
      >
        Height
      </label>
      <input
        id="form-height"
        name="height"
        placeholder="Height"
        type="text"
        value=""
      />
      <label
        for="form-backgroundColor"
      >
        Color
      </label>
      <input
        id="form-backgroundColor"
        name="backgroundColor"
        placeholder="Color"
        type="text"
        value=""
      />
      <button>
        Add Box
      </button>
    </form>
  </div>
  <div
    class="BoxList"
  />
</DocumentFragment>
`;
